tree/treeplayer/src/TDFUtils.cxx

```diff
@@ -114,23 +114,31 @@ std::string TypeID2TypeName(const std::type_info &id)
 
 /// Return a string containing the type of the given branch. Works both with real TTree branches and with temporary
 /// column created by Define. Returns an empty string if type name deduction fails.
+/// Note that for fixed- or variable-sized c-style arrays as well std::vector<T>, the returned type name will be TVec<T>
 std::string
 ColumnName2ColumnTypeName(const std::string &colName, TTree *tree, TCustomColumnBase *tmpBranch, TDataSource *ds)
 {
+   std::string colType;
+
    // if this is a TDataSource column, we just ask the type name to the data-source
-   if (ds && ds->HasColumn(colName)) {
-      return ds->GetTypeName(colName);
-   }
+   if (ds && ds->HasColumn(colName))
+      colType = ds->GetTypeName(colName);
 
-   TBranch *branch = nullptr;
-   if (tree)
-      branch = tree->GetBranch(colName.c_str());
+   TBranch *branch = tree ? tree->GetBranch(colName.c_str()) : nullptr;
    if (branch) {
       // this must be a real TTree branch
       static const TClassRef tbranchelRef("TBranchElement");
       if (branch->InheritsFrom(tbranchelRef)) {
          // this branch is not a fundamental type, we can ask for the class name
-         return static_cast<TBranchElement *>(branch)->GetClassName();
+         std::string classname(static_cast<TBranchElement *>(branch)->GetClassName());
+         if (classname.compare(0, 7, "vector<") == 0) {
+            // column type is "vector<ValueType>", we read it as "TVec<ValueType>"
+            // value type is the classname.size() - 8 chars after the 7th character in "vector<ValueType>"
+            auto valueType = classname.substr(7, classname.size() - 8);
+            colType = "ROOT::Experimental::VecOps::TVec<" + valueType + ">";
```

Comment:
We build this string over and over: maybe we can add a helper function?
Another detail: we may perhaps avoid copies taking a TVec by reference? In any case the new code allows us to do so!

Reply:
what do you mean? the string is built ~once per call and it should satisfy all requirements for the c++17 guaranteed NRVO afaik.

The reference after the type name is added when appropriate (always, in the case of TVecs) by other methods, e.g. `JitTransormation`

Reply:
p.s. anyway we should look into avoiding calling `ColumnName2ColumnTypeName` for each column in each transformation/action -- we could build a TDataFrame-wide register that maps columns to types and only perform the lookup once per column per program per TDataFrame graph.

Reply:
nothing so sophisticated.. Just to centralise the name of the type rather than writing it 3 times, that's all :)

Reply:
p.s. and invoke it as many times as needed :)

Reply:
ah I understand now! yes, certainly